iceberg-rust/src/table/transaction/append.rs

```diff
@@ -127,70 +129,122 @@ pub(crate) fn select_manifest_partitioned(
 
-        let Some((selected_row_count, selected_manifest)) = &selected_state else {
-            selected_state = Some((row_count, manifest));
-            continue;
-        };
-
-        // If the file doesn't have any rows, we select it
-        let Some(row_count) = row_count else {
-            selected_state = Some((row_count, manifest));
-            continue;
-        };
-
-        if selected_row_count.is_some_and(|x| x > row_count) {
-            manifest_list_writer.append_ser(selected_manifest)?;
-            selected_state = Some((Some(row_count), manifest));
-            continue;
-        } else {
-            manifest_list_writer.append_ser(manifest)?;
-            continue;
+        match manifest.content {
+            iceberg_rust_spec::manifest_list::Content::Data => {
+                let Some((selected_row_count, selected_manifest)) = &selected_data_state else {
+                    selected_data_state = Some((row_count, manifest));
+                    continue;
+                };
+
+                // If the file doesn't have any rows, we select it
+                let Some(row_count) = row_count else {
+                    selected_data_state = Some((row_count, manifest));
+                    continue;
+                };
+
+                if selected_row_count.is_some_and(|x| x > row_count) {
+                    manifest_list_writer.append_ser(selected_manifest)?;
+                    selected_data_state = Some((Some(row_count), manifest));
+                    continue;
+                } else {
+                    manifest_list_writer.append_ser(manifest)?;
+                    continue;
+                }
+            }
+            iceberg_rust_spec::manifest_list::Content::Deletes => {
+                let Some((selected_row_count, selected_manifest)) = &selected_delete_state else {
+                    selected_delete_state = Some((row_count, manifest));
+                    continue;
+                };
+
+                // If the file doesn't have any rows, we select it
+                let Some(row_count) = row_count else {
+                    selected_delete_state = Some((row_count, manifest));
+                    continue;
+                };
+
+                if selected_row_count.is_some_and(|x| x > row_count) {
+                    manifest_list_writer.append_ser(selected_manifest)?;
+                    selected_delete_state = Some((Some(row_count), manifest));
+                    continue;
+                } else {
+                    manifest_list_writer.append_ser(manifest)?;
+                    continue;
+                }
+            }
         }
     }
-    selected_state
-        .map(|(_, entry)| SelectedManifest {
-            manifest: entry,
-            file_count_all_entries,
-        })
-        .ok_or(Error::NotFound("Manifest for insert".to_owned()))
+    let (_, data_manifest) =
+        selected_data_state.ok_or(Error::NotFound("Manifest for insert".to_owned()))?;
```

Comment:
Maybe this shouldn't be an error but instead, the entire `Option<ManifestListEntry>` should be returned, since the caller (`ManifestListWriter::from_existing`) knows how to handle that correctly.

Granted I think this scenario can only happen if we have an empty/new table and we wont to append some deletes to it, or some such edge case.